File: src/main/java/me/taubsie/carrylogs/application/command/commands/ConfigCommand.java
package me.taubsie.carrylogs.application.command.commands;

import me.taubsie.carrylogs.application.classes.ServerData;
import me.taubsie.carrylogs.application.classes.ServerProperty;
import me.taubsie.carrylogs.application.classes.ServerPropertyType;
import me.taubsie.carrylogs.application.command.Command;
import me.taubsie.carrylogs.application.command.CommandParameters;
import me.taubsie.carrylogs.application.enums.EmbedColor;
import me.taubsie.carrylogs.application.exceptions.CommandExecutionException;
import me.taubsie.carrylogs.application.exceptions.InvalidOptionException;
import me.taubsie.carrylogs.application.service.ApplicationService;
import me.taubsie.carrylogs.application.service.ServerService;
import org.javacord.api.entity.channel.ChannelType;
import org.javacord.api.entity.permission.PermissionType;
import org.javacord.api.event.interaction.SlashCommandCreateEvent;
import org.javacord.api.interaction.*;

import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@CommandParameters(name = "config", description = "Edits the config for the server.", enabledForPermissions =
        {PermissionType.ADMINISTRATOR})
public class ConfigCommand extends Command {
    @Override
    protected void executeCommand(SlashCommandCreateEvent slashCommandCreateEvent) {
        boolean get = slashCommandCreateEvent.getSlashCommandInteraction().getOptionByName("get").isPresent();

        Optional<ServerProperty> property = slashCommandCreateEvent
                .getSlashCommandInteraction()
                .getOptionByIndex(0)
                .map(option -> getStringOption(option, "property"))
                .flatMap(ServerProperty::getPropertyByName);

        if(property.isEmpty()) {
            throw new InvalidOptionException(
                    "property",
                    "Please use one of the following: "
                            + Arrays.stream(ServerProperty.values()).map(ServerProperty::getName).collect(Collectors.joining(", "))
            );
        }

        if(get) {
            getConfig(property.get());
        } else {
            //TODO maybe enable if user is bot owner
            if(!property.get().isEnabled(getServer().getId())) {
                throw new InvalidOptionException("property", "Property is disabled on this server.");
            }

            Optional<String> value = slashCommandCreateEvent
                    .getSlashCommandInteraction()
                    .getOptionByIndex(0)
                    .map(option -> {
                        if(property.get().getPropertyType() == ServerPropertyType.ROLE) {
                            return getRoleOption(option, "value").getIdAsString();
                        }

                        if(property.get().getPropertyType() == ServerPropertyType.CHANNEL) {
                            return getChannelOption(option, "value").getIdAsString();
                        }

                        if(property.get().getPropertyType() == ServerPropertyType.NUMBER) {
                            return String.valueOf(getNumberOption(option, "value"));
                        }

                        return getStringOption(option, "value");
                    })
                    .map(s -> s.replace("\\n", "\n"));

            if(value.isEmpty()) {
                throw new InvalidOptionException("value", "Please enter a new value.");
            }

            setConfig(property.get(), value.get());
        }
    }

    private void setConfig(ServerProperty property, String value) {
        Optional<String> previousValue = ServerService.getInstance().getActualServerProperty(getServer().getId(),
                property);

        String oldValue = previousValue.isPresent()
                ? property.getPropertyType().applyPropertyType(previousValue.get())
                : "None was set.";

        Optional<ServerData> serverData = ServerService.getInstance().getServerData(getServer().getId());

        if(serverData.isEmpty()) {
            throw new CommandExecutionException() {
                @Override
                public String getMessage() {
                    return "Couldn't load the server data from storage.";
                }
            };
        }

        serverData.get().setConfig(property, value);

        respond(ApplicationService.getInstance()
                .getEmbed()
                .setColor(EmbedColor.POSITIVE.getColor())
                .setDescription("Changed the value of `" + property.getName() + "`.")
                .addInlineField("Old value", oldValue)
                .addInlineField("New value", property.getPropertyType().applyPropertyType(value)));
    }

    private void getConfig(ServerProperty property) {
        Optional<String> value = ServerService.getInstance().getActualServerProperty(getServer().getId(), property);

        if(value.isEmpty()) {
            respondEphemeral(ApplicationService.getInstance()
                    .getEmbed()
                    .setColor(EmbedColor.NEGATIVE.getColor())
                    .setDescription("No value for `" + property.getName() + "` is set.")
                    .addInlineField("Option enabled", String.valueOf(property.isEnabled(getServer().getId()))));
            return;
        }

        respond(ApplicationService.getInstance()
                .getEmbed()
                .setColor(EmbedColor.INFORMATION.getColor())
                .setDescription("Loaded the value of `" + property.getName() + "`.")
                .addInlineField("Current value", property.getPropertyType().applyPropertyType(value.get()))
                .addInlineField("Option enabled", String.valueOf(property.isEnabled(getServer().getId()))));
    }

    @Override
    public List<SlashCommandOption> getSlashCommandOptions() {
        SlashCommandOption allPropertiesOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.STRING)
                .setName("property")
                .setDescription("The property to choose.")
                .setChoices(Arrays.stream(ServerProperty.values())
                        .map(ServerProperty::getName)
                        .map(s -> new SlashCommandOptionChoiceBuilder().setName(s).setValue(s).build())
                        .toList())
                .setRequired(true)
                .build();

        //TODO make autocompletable and maybe add options? idk
        SlashCommandOption propertyOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.STRING)
                .setName("property")
                .setDescription("The property to choose.")
                .setChoices(Arrays.stream(ServerProperty.values())
                        .filter(serverProperty -> serverProperty.getPropertyType() == ServerPropertyType.STRING)
                        .map(ServerProperty::getName)
                        .map(s -> new SlashCommandOptionChoiceBuilder().setName(s).setValue(s).build())
                        .toList())
                .setRequired(true)
                .build();

        SlashCommandOption channelPropertyOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.STRING)
                .setName("property")
                .setDescription("The channel property to choose.")
                .setChoices(Arrays.stream(ServerProperty.values())
                        .filter(serverProperty -> serverProperty.getPropertyType() == ServerPropertyType.CHANNEL)
                        .map(ServerProperty::getName)
                        .map(s -> new SlashCommandOptionChoiceBuilder().setName(s).setValue(s).build())
                        .toList())
                .setRequired(true)
                .build();

        SlashCommandOption rolePropertyOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.STRING)
                .setName("property")
                .setDescription("The role property to choose.")
                .setChoices(Arrays.stream(ServerProperty.values())
                        .filter(serverProperty -> serverProperty.getPropertyType() == ServerPropertyType.ROLE)
                        .map(ServerProperty::getName)
                        .map(s -> new SlashCommandOptionChoiceBuilder().setName(s).setValue(s).build())
                        .toList())
                .setRequired(true)
                .build();

        SlashCommandOption valueOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.STRING)
                .setName("value")
                .setDescription("What you want to set the property to.")
                .setRequired(true)
                .build();

        SlashCommandOption channelValueOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.CHANNEL)
                .setName("value")
                .setDescription("What you want to set the property to.")
                .setRequired(true)
                .setChannelTypes(List.of(ChannelType.CHANNEL_CATEGORY, ChannelType.SERVER_TEXT_CHANNEL))
                .build();

        SlashCommandOption roleValueOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.ROLE)
                .setName("value")
                .setDescription("What you want to set the property to.")
                .setRequired(true)
                .build();

        SlashCommandOption getCommand = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.SUB_COMMAND)
                .setName("get")
                .setDescription("Shows you the current config for the server.")
                .setOptions(List.of(allPropertiesOption))
                .build();

        SlashCommandOption setCommand = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.SUB_COMMAND)
                .setName("set")
                .setDescription("Sets the config.")
                .setOptions(List.of(propertyOption, valueOption))
                .build();

        SlashCommandOption setChannelCommand = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.SUB_COMMAND)
                .setName("set-channel")
                .setDescription("Sets the config for channels.")
                .setOptions(List.of(channelPropertyOption, channelValueOption))
                .build();

        SlashCommandOption setRoleCommand = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.SUB_COMMAND)
                .setName("set-role")
                .setDescription("Sets the config for roles.")
                .setOptions(List.of(rolePropertyOption, roleValueOption))
                .build();

        return Arrays.asList(getCommand, setCommand, setChannelCommand, setRoleCommand);
    }
}

File: src/main/kotlin/net/dungeonhub/application/commands/ConfigCommand.kt
package net.dungeonhub.application.commands

import dev.kord.common.entity.ChannelType
import dev.kord.common.entity.Permission
import dev.kord.common.entity.Permissions
import dev.kord.rest.builder.message.EmbedBuilder
import dev.kordex.core.commands.Arguments
import dev.kordex.core.commands.application.slash.converters.impl.stringChoice
import dev.kordex.core.commands.application.slash.group
import dev.kordex.core.commands.application.slash.publicSubCommand
import dev.kordex.core.commands.converters.impl.*
import dev.kordex.core.extensions.Extension
import dev.kordex.core.extensions.publicSlashCommand
import dev.kordex.core.i18n.toKey
import net.dungeonhub.application.enums.EmbedColor
import net.dungeonhub.application.enums.ServerProperty
import net.dungeonhub.application.enums.ServerPropertyType
import net.dungeonhub.application.exceptions.CommandExecutionException
import net.dungeonhub.application.exceptions.InvalidOptionException
import net.dungeonhub.application.loader.LoadExtension
import net.dungeonhub.application.service.ApplicationService
import net.dungeonhub.application.service.ServerService
import net.dungeonhub.i18n.Translations.Command.Config
import java.util.stream.Collectors

/**
 * The config command allows you to edit the config for the server.
 * This includes setting and getting values for the server.
 * The command is only available for users with the Administrator permission and is also only available in guilds.
 * The command has the following subcommands:
 * - `get`: Shows you the current config for the server.
 * - `set string`: Sets a string config value.
 * - `set number`: Sets a number config value.
 * - `set boolean`: Sets a boolean config value.
 * - `set channel`: Sets a channel config value.
 * - `set category`: Sets a category config value.
 * - `set role`: Sets a role config value.
 * - `set null`: Sets a property to null (resetting it).
 * @see ServerProperty
 */
@LoadExtension
class ConfigCommand : Extension() {
    override val name = "config-command"

    override suspend fun setup() {
        publicSlashCommand {
            name = Config.name
            description = Config.description
            defaultMemberPermissions = Permissions(Permission.Administrator)
            allowInDms = false

            publicSubCommand(::GetArguments) {
                name = "get".toKey()
                description = "Shows you the current config for the server.".toKey()

                action {
                    respond {
                        val guildId = guild!!.id.value.toLong()

                        val property = arguments.getProperty(guildId)
                        val value = ServerService.getActualServerProperty(guildId, property).orElse(null)

                        if (value == null) {
                            val embed = ApplicationService.embed
                            embed.color = EmbedColor.Negative.color
                            embed.description = "No value for `${property.name}` is set."

                            if (!property.isEnabled(guildId)) {
                                embed.field("Option enabled", true) {
                                    property.isEnabled(guildId).toString()
                                }
                            }

                            embeds = mutableListOf(embed)
                            return@respond
                        }

                        val embed = ApplicationService.embed
                        embed.color = EmbedColor.Information.color
                        embed.description = "Loaded the value of `${property.name}`."
                        embed.field("Current value", true) { property.propertyType.applyPropertyType(value) }

                        if (!property.isEnabled(guildId)) {
                            embed.field("Option enabled", true) {
                                property.isEnabled(guildId).toString()
                            }
                        }

                        embeds = mutableListOf(embed)
                    }
                }
            }

            group("set".toKey()) {
                description = "Sets a new config property".toKey()

                publicSubCommand(::SetStringArguments) {
                    name = "string".toKey()
                    description = "Sets a config value.".toKey()

                    action {
                        respond {
                            val guildId = guild!!.id.value.toLong()

                            embeds = mutableListOf(setConfig(arguments.getProperty(guildId), arguments.value, guildId))
                        }
                    }
                }

                publicSubCommand(::SetNumberArguments) {
                    name = "number".toKey()
                    description = "Sets a config value.".toKey()

                    action {
                        respond {
                            val guildId = guild!!.id.value.toLong()

                            embeds = mutableListOf(
                                setConfig(
                                    arguments.getProperty(guildId),
                                    arguments.value.toString(),
                                    guildId
                                )
                            )
                        }
                    }
                }

                publicSubCommand(::SetBooleanArguments) {
                    name = "boolean".toKey()
                    description = "Sets a config value.".toKey()

                    action {
                        respond {
                            val guildId = guild!!.id.value.toLong()

                            embeds = mutableListOf(
                                setConfig(
                                    arguments.getProperty(guildId),
                                    arguments.value.toString(),
                                    guildId
                                )
                            )
                        }
                    }
                }

                publicSubCommand(::SetChannelArguments) {
                    name = "channel".toKey()
                    description = "Sets a config value.".toKey()

                    action {
                        respond {
                            val guildId = guild!!.id.value.toLong()

                            embeds = mutableListOf(
                                setConfig(
                                    arguments.getProperty(guildId),
                                    arguments.value.id.value.toString(),
                                    guildId
                                )
                            )
                        }
                    }
                }

                publicSubCommand(::SetCategoryArguments) {
                    name = "category".toKey()
                    description = "Sets a config value.".toKey()

                    action {
                        respond {
                            val guildId = guild!!.id.value.toLong()

                            embeds = mutableListOf(
                                setConfig(
                                    arguments.getProperty(guildId),
                                    arguments.value.id.toString(),
                                    guildId
                                )
                            )
                        }
                    }
                }

                publicSubCommand(::SetRoleArguments) {
                    name = "role".toKey()
                    description = "Sets a config value.".toKey()

                    action {
                        respond {
                            val guildId = guild!!.id.value.toLong()

                            embeds = mutableListOf(
                                setConfig(
                                    arguments.getProperty(guildId),
                                    arguments.value.id.toString(),
                                    guildId
                                )
                            )
                        }
                    }
                }

                publicSubCommand(::GetArguments) {
                    name = "null".toKey()
                    description = "Sets a property to null (resetting it).".toKey()

                    action {
                        respond {
                            val guildId = guild!!.id.value.toLong()

                            val property = arguments.getProperty(guildId)

                            if (!property.isEnabled(guildId)) {
                                throw InvalidOptionException("property", "This property is disabled on this server.")
                            }

                            val oldValue = ServerService.getActualServerProperty(guildId, property)
                                .map { property.propertyType.applyPropertyType(it) }
                                .orElse("None was set.")

                            val serverData = ServerService.getServerData(guildId)

                            if (serverData.isEmpty) {
                                throw CommandExecutionException("Couldn't load the server data from storage.")
                            }

                            serverData.get().setConfig(property, null)

                            val embed = ApplicationService.embed
                            embed.color = EmbedColor.Positive.color
                            embed.description = "Changed the value of `${property.name}`."
                            embed.field("Old value", true) {
                                oldValue
                            }
                            embed.field("New value", true) {
                                "None"
                            }

                            embeds = mutableListOf(embed)
                        }
                    }
                }
            }
        }
    }

    fun setConfig(property: ServerProperty, newValue: String, guildId: Long): EmbedBuilder {
        if (!property.isEnabled(guildId)) {
            throw InvalidOptionException("property", "This property is disabled on this server.")
        }

        val value = newValue.replace("\\n", "\n")

        if (value.isBlank()) {
            throw InvalidOptionException("value", "Please enter a new value.")
        }

        val oldValue = ServerService.getActualServerProperty(guildId, property)
            .map { property.propertyType.applyPropertyType(it) }
            .orElse("None was set.")

        val serverData = ServerService.getServerData(guildId)

        if (serverData.isEmpty) {
            throw CommandExecutionException("Couldn't load the server data from storage.")
        }

        serverData.get().setConfig(property, value)

        val embed = ApplicationService.embed
        embed.color = EmbedColor.Positive.color
        embed.description = "Changed the value of `${property.name}`."
        embed.field("Old value", true) {
            oldValue
        }
        embed.field("New value", true) {
            property.propertyType.applyPropertyType(value)
        }

        return embed
    }

    inner class GetArguments : Arguments() {
        val property by stringChoice {
            name = "property".toKey()
            description = "The property to choose.".toKey()
            choices = ServerProperty.entries.map { it.name.toKey() to it.name }.toMap().toMutableMap()
        }

        fun getProperty(serverId: Long): ServerProperty {
            val property = ServerProperty.getPropertyByName(property).orElse(null)
                ?: throw InvalidOptionException(
                    "property",
                    "Please use one of the following: ${
                        ServerProperty.entries
                            .filter { it.isEnabled(serverId) }
                            .map { it.name }.stream()
                            .collect(Collectors.joining(", "))
                    }"
                )

            return property
        }
    }

    inner class SetStringArguments : Arguments() {
        val property by stringChoice {
            name = "property".toKey()
            description = "The string property to choose.".toKey()
            choices = ServerProperty.entries
                .filter { it.propertyType == ServerPropertyType.STRING }
                .map { it.name.toKey() to it.name }
                .toMap().toMutableMap()
        }

        val value by string {
            name = "value".toKey()
            description = "What you want to set the property to.".toKey()
        }

        fun getProperty(serverId: Long): ServerProperty {
            val property = ServerProperty.getPropertyByName(property).orElse(null)
                ?: throw InvalidOptionException(
                    "property",
                    "Please use one of the following: ${
                        ServerProperty.entries
                            .filter { it.isEnabled(serverId) }
                            .map { it.name }.stream()
                            .collect(Collectors.joining(", "))
                    }"
                )

            return property
        }
    }

    inner class SetNumberArguments : Arguments() {
        val property by stringChoice {
            name = "property".toKey()
            description = "The number property to choose.".toKey()
            choices = ServerProperty.entries
                .filter { it.propertyType == ServerPropertyType.NUMBER }
                .map { it.name.toKey() to it.name }
                .toMap().toMutableMap()
        }

        val value by long {
            name = "value".toKey()
            description = "What you want to set the property to.".toKey()
        }

        fun getProperty(serverId: Long): ServerProperty {
            val property = ServerProperty.getPropertyByName(property).orElse(null)
                ?: throw InvalidOptionException(
                    "property",
                    "Please use one of the following: ${
                        ServerProperty.entries
                            .filter { it.isEnabled(serverId) }
                            .map { it.name }.stream()
                            .collect(Collectors.joining(", "))
                    }"
                )

            return property
        }
    }

    inner class SetBooleanArguments : Arguments() {
        val property by stringChoice {
            name = "property".toKey()
            description = "The boolean property to choose.".toKey()
            choices = ServerProperty.entries
                .filter { it.propertyType == ServerPropertyType.BOOLEAN }
                .map { it.name.toKey() to it.name }
                .toMap().toMutableMap()
        }

        val value by boolean {
            name = "value".toKey()
            description = "What you want to set the property to.".toKey()
        }

        fun getProperty(serverId: Long): ServerProperty {
            val property = ServerProperty.getPropertyByName(property).orElse(null)
                ?: throw InvalidOptionException(
                    "property",
                    "Please use one of the following: ${
                        ServerProperty.entries
                            .filter { it.isEnabled(serverId) }
                            .map { it.name }.stream()
                            .collect(Collectors.joining(", "))
                    }"
                )

            return property
        }
    }

    inner class SetChannelArguments : Arguments() {
        val property by stringChoice {
            name = "property".toKey()
            description = "The channel property to choose.".toKey()
            choices = ServerProperty.entries
                .filter { it.propertyType == ServerPropertyType.CHANNEL }
                .map { it.name.toKey() to it.name }
                .toMap().toMutableMap()
        }

        val value by channel {
            name = "value".toKey()
            description = "What you want to set the property to.".toKey()
            requiredChannelTypes = mutableSetOf(
                ChannelType.GuildText
            )
        }

        fun getProperty(serverId: Long): ServerProperty {
            val property = ServerProperty.getPropertyByName(property).orElse(null)
                ?: throw InvalidOptionException(
                    "property",
                    "Please use one of the following: ${
                        ServerProperty.entries
                            .filter { it.isEnabled(serverId) }
                            .map { it.name }.stream()
                            .collect(Collectors.joining(", "))
                    }"
                )

            return property
        }
    }

    inner class SetCategoryArguments : Arguments() {
        val property by stringChoice {
            name = "property".toKey()
            description = "The category property to choose.".toKey()
            choices = ServerProperty.entries
                .filter { it.propertyType == ServerPropertyType.CATEGORY }
                .map { it.name.toKey() to it.name }
                .toMap().toMutableMap()
        }

        val value by channel {
            name = "value".toKey()
            description = "What you want to set the property to.".toKey()
            requiredChannelTypes = mutableSetOf(
                ChannelType.GuildCategory
            )
        }

        fun getProperty(serverId: Long): ServerProperty {
            val property = ServerProperty.getPropertyByName(property).orElse(null)
                ?: throw InvalidOptionException(
                    "property",
                    "Please use one of the following: ${
                        ServerProperty.entries
                            .filter { it.isEnabled(serverId) }
                            .map { it.name }.stream()
                            .collect(Collectors.joining(", "))
                    }"
                )

            return property
        }
    }

    inner class SetRoleArguments : Arguments() {
        val property by stringChoice {
            name = "property".toKey()
            description = "The role property to choose.".toKey()
            choices = ServerProperty.entries
                .filter { it.propertyType == ServerPropertyType.ROLE }
                .map { it.name.toKey() to it.name }
                .toMap().toMutableMap()
        }

        val value by role {
            name = "value".toKey()
            description = "What you want to set the property to.".toKey()
        }

        fun getProperty(serverId: Long): ServerProperty {
            val property = ServerProperty.getPropertyByName(property).orElse(null)
                ?: throw InvalidOptionException(
                    "property",
                    "Please use one of the following: ${
                        ServerProperty.entries
                            .filter { it.isEnabled(serverId) }
                            .map { it.name }.stream()
                            .collect(Collectors.joining(", "))
                    }"
                )

            return property
        }
    }
}